src/gui/creds/flow2auth.cpp

```diff
@@ -0,0 +1,164 @@
+/*
+ * Copyright (C) by Olivier Goffart <ogoffart@woboq.com>
+ * Copyright (C) by Michael Schuster <michael@nextcloud.com>
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
+ * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License
+ * for more details.
+ */
+
+#include <QDesktopServices>
+#include <QTimer>
+#include <QBuffer>
+#include "account.h"
+#include "creds/flow2auth.h"
+#include <QJsonObject>
+#include <QJsonDocument>
+#include "theme.h"
+#include "networkjobs.h"
+#include "configfile.h"
+
+namespace OCC {
+
+Q_LOGGING_CATEGORY(lcFlow2auth, "nextcloud.sync.credentials.flow2auth", QtInfoMsg)
+
+Flow2Auth::~Flow2Auth()
+{
+}
+
+void Flow2Auth::start()
+{
+	// Note: All startup code is in openBrowser() to allow reinitiate a new request with
+	//       fresh tokens. Opening the same pollEndpoint link twice triggers an expiration
+	//       message by the server (security, intended design).
+    openBrowser();
+}
+
+QUrl Flow2Auth::authorisationLink() const
+{
+    return _loginUrl;
+}
+
+void Flow2Auth::openBrowser()
+{
+    _pollTimer.stop();
+
+    // Step 1: Initiate a login, do an anonymous POST request
+    QUrl url = Utility::concatUrlPath(_account->url().toString(), QLatin1String("/index.php/login/v2"));
+
+    auto job = _account->sendRequest("POST", url);
+    job->setTimeout(qMin(30 * 1000ll, job->timeoutMsec()));
+
+    QObject::connect(job, &SimpleNetworkJob::finishedSignal, this, [this](QNetworkReply *reply) {
+        auto jsonData = reply->readAll();
+        QJsonParseError jsonParseError;
+        QJsonObject json = QJsonDocument::fromJson(jsonData, &jsonParseError).object();
+
+        QString pollToken = json.value("poll").toObject().value("token").toString();
+        QString pollEndpoint = json.value("poll").toObject().value("endpoint").toString();
+        QUrl loginUrl = json["login"].toString();
+
+        if (reply->error() != QNetworkReply::NoError || jsonParseError.error != QJsonParseError::NoError
```

Comment:
Shouldn't this be moved up ? So we do not even try to parse things if there is an error?

Reply:
Then we would have to duplicate more from the code than we're saving ;-)